File: src/main/java/state/CommandRunner.java
package state;

import antlr.CommandBaseVisitor;
import antlr.CommandParser;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.misc.NotNull;

import java.util.List;
import java.util.StringJoiner;
import java.util.stream.Collectors;

public class CommandRunner extends CommandBaseVisitor<Void> {
    private final Game gameRoom;
    private final Player issuingPlayer;

    public CommandRunner(Game gameRoom, Player issuingPlayer) {
        this.gameRoom = gameRoom;
        this.issuingPlayer = issuingPlayer;
    }

    private String joinLongName(@NotNull CommandParser.LongNameContext ctx) {
        StringJoiner stringJoiner = new StringJoiner(" ");
        ctx.Identifier().stream().forEach(i -> stringJoiner.add(i.getText()));
        return stringJoiner.toString();
    }

    @Override
    public Void visitRelativeMovement(@NotNull CommandParser.RelativeMovementContext ctx) {
        String direction = ctx.direction().getText();
        switch (direction) {
            case "n": direction = "north"; break;
            case "s": direction = "south"; break;
            case "e": direction = "east"; break;
            case "w": direction = "west"; break;
            default:
        }
        gameRoom.move(issuingPlayer, direction);
        System.out.println("going to the " + direction);
        return super.visitRelativeMovement(ctx);
    }

    public static void execute(Game gameRoom, Player issuingPlayer, ParseTree tree) {
        new CommandRunner(gameRoom, issuingPlayer).visit(tree);
    }
}

File: src/main/scala/server/CommandRunner.scala
package server

import antlr.CommandParser._
import antlr.{CommandBaseVisitor, CommandParser}
import org.antlr.v4.runtime.tree.ParseTree
import state.{Game, Player}

import scala.collection.JavaConverters._

class CommandRunner(gameRoom: Game, issuingPlayer: Player) extends CommandBaseVisitor[Void] {
  override def visitRelativeMovement(ctx: RelativeMovementContext): Void = {
    gameRoom.move(issuingPlayer, ctx.direction().getText match {
      case "n" => "north";
      case "s" => "south";
      case "e" => "east";
      case "w" => "west";
      case dir => dir;
    })
    super.visitRelativeMovement(ctx)
  }

  override def visitInteraction(ctx: InteractionContext): Void = {
    val longName = joinLongName(ctx.longName())

    if (ctx.Drop() != null) {
      issuingPlayer.dropItem(longName, if (ctx.Number() != null) Some(ctx.Number().getText.toInt) else None)
    } else if (ctx.Eat() != null) {
      gameRoom.eat(issuingPlayer, longName);
    } else if (ctx.Take() != null || ctx.Pick() != null) {
      gameRoom.take(issuingPlayer, longName);
    }

    super.visitInteraction(ctx)
  }


  override def visitAttack(ctx: AttackContext): Void = {
    val target = joinLongName(ctx.longName(0))
    val weapon = Option(joinLongName(ctx.longName(1)))

    gameRoom.attack(issuingPlayer, target, weapon)

    super.visitAttack(ctx)
  }

  override def visitInspection(ctx: InspectionContext): Void = {
    val longName = joinLongName(ctx.longName())

    if (ctx.Look() != null || ctx.Examine() != null || ctx.Inspect() != null) {
      gameRoom.lookAt(issuingPlayer, longName);
    } else if (ctx.Open() != null) {
      gameRoom.open(issuingPlayer, longName);
    }

    super.visitInspection(ctx)
  }

  override def visitSpecificMovement(ctx: SpecificMovementContext): Void = {
    gameRoom.move(issuingPlayer, joinLongName(ctx.longName()))
    super.visitSpecificMovement(ctx)
  }

  private def joinLongName(ctx: CommandParser.LongNameContext): String = {
    ctx.Identifier().asScala.map(_.getText).mkString(" ")
  }

  override def visitCheckHealth(ctx: CheckHealthContext): Void = {
    issuingPlayer.notify("You have " + issuingPlayer.health + " hit points left.")
    super.visitCheckHealth(ctx)
  }

  override def visitCheckStatus(ctx: CheckStatusContext): Void = {
    issuingPlayer.notify("You have " + issuingPlayer.health + " hit points left.")
    super.visitCheckStatus(ctx)
  }

  override def visitInventory(ctx: InventoryContext): Void = {
    if (issuingPlayer.inventory.isEmpty) issuingPlayer.notify("Your inventory is empty.")
    else issuingPlayer.inventory.zipWithIndex.foreach({
      case (item, i) => issuingPlayer.notify("(%d) %s: %s".format(i + 1, item.name, item.description))
    })
    super.visitInventory(ctx)
  }
}

object CommandRunner {
  def execute(gameRoom: Game, issuingPlayer: Player, tree: ParseTree) = {
    new CommandRunner(gameRoom, issuingPlayer).visit(tree)
  }
}

